### ISSUE TITLE
Roadmap for Model Parsing Upgrade

### ISSUE THREAD

**Issue by scidom** [JuliaStats/Lora.jl#71]
@fredo-dedup I was thinking that the next massive improvement for our package would be to generalize some aspects of model specification, namely:

* allow partial updates of subsets of model parameters,
* allow updates of correlated parameters,
* enable hierarchical models.
* this among else would tackle Gibbs sampling as a subcase, on which @spencerlyon2 has made some work.

I haven't touched model specification at all until today. Does the relevant code reside in src/parsers/modelparser.jl? Have you made any progress in any other branch or repo, to make sure we avoid duplication of work? I was thinking of using the Julia Graphs package for model parsing (still in an imperative fashion, I don't plan to go for declarative programming), what do you think, would that be a good direction?

**Comment by fredo-dedup** [comment]
Hi @scidom , I made some progress in integrating the new ReverseDiffSource in Lora. Along the way, I made some improvements to ReverseDiffSource and found a few bugs to correct (I have still a tough one to solve).
However I am now a bit stuck because I can't make Lora work in its current state (apparently around the MCJobs definition). I could change a few things to make it work but in the process I might also break the structure you were trying to build because the intention is not entirely clear to me. Could you provide some more information on how MCJobs is supposed to work ?
I was also struck on how much the package has grown complicated with time ! Notably the many nested types. What do you think of trying to simplify things as much as possible to make the package more accessible for potential contributions (including mine ! ). This could be tackled by posting separate issues. However, if you prefer to continue the rewrite to reach a stable state and fully flesh out your ideas before, I can understand too.

**Comment by scidom** [comment]
Hi @fredo-dedup, I will try to provide a succinct recap on the 2 major recent additions to the package, both of which relate to the new `MCJob` interface.

To start with, the main idea has been that a common structure underlines the running of any sampler. You may think of this at an abstract simplistic level as a loop iterating over the sampler "iterator". The only distinguishing feature of each sampler then is how the iteration function is defined. Thus each sampler defines its own `iterate!` method.

The second refactoring relates to the fact that each job type mantains its own `send` and `receive` fields. These methods essentially relate to how the `iterate!()` method communicates with the outer loop. This way, it becomes possible to switch between a task-based or a plain job which does not use `Task` for controlling the flow.

If you wish, I could provide more techinical details, this was a high-level description.

In my humble opinion, a piece of software intended to offer broad functionality naturally rises to a level of complexity with time. My opinion is that the solution does not lie on sacrifising features but on providing documentation. As an extreme example, if LLVM was simplified, that would unavoidably change its scope. Instead, the LLVM documentation expanded over the last 2 years to address the issue of software engineering complexity and become more developer-friendly.

An other underlying issue is in my opinion the lack of coordination and communication, which albeit happening unintentionally, it has led to negative effects in coding development. I sincerely don't know what is the solution to this problem, esp given our limited time.

From my side, I have been doing some background work towards 2 different directions, one being use of graphical models for model specification and the second one being recursive implementation of Monte Carlo methodology using a functional programming approach; as far as the latter goes, it is possible to define for instance Gibbs sampling by employing a functional programming paradigm using tail recursion, which works in general (although I don't know how well it will work out at the end in Julia because I don't know how much work has been put into enabling effective tail recursion in Julia while simultaneously ensuring that no stack overflow occurs). I deviated a bit at this point with the intention of unfolding and sharing my thoughts and long-shot roadmap.

Given the intrinsic complexity of the situation, your suggestion may be the best; it may be the best option that you revert the recent developments in Lora so that we avoid hindering contribution, if this is the heart of the issue. I can put up my own thoughts and directions into a separate package in the future, which seems to naturally align with the recent independent MCMC Julia developments. I am open to counter-suggestions and discussion, publicly encouraging it.

**Comment by fredo-dedup** [comment]
I can see a little bit better now, thanks for your explanations. No need to
break the work you have been doing. I'll try to contribute in the 'devel'
branch to restore the functionality of run() and validate the result with
you. Hopefully we'll be able to proceed that way for the remaining steps.
You are right that going through architectural changes like that are
difficult with our limited availability. It is probably best that I follow
your ideas instead causing perturbations  when you're half way implementing
them !
Le 25 janv. 2015 20:21, "Theodore Papamarkou" <notifications@github.com> a
écrit :

**Comment by scidom** [comment]
Retrospectively, after reflecing on the recent developments of our package, I came to the conclusion that I had made a strategic mistake; I have tried to bring in all these reforms as soon as possible instead of letting my work mature in a devel branch. I had been undergoing an internal debate as to whether it would be better to push forward the changes asap or use a devel branch. Limited time has perhaps obscured my decision-making capacity leading me to think that submitting it all to the master would be attainable ambition.

In any case, what I need to do is to carry on with the package documentation, which will make it much easier to communicate and interact.

It is complicated to progress when we don't have enough time to fully commit to this project, mostly because the natural obssession that is developed via coding can't find enough time to flourish. I will try to push myself a bit more to carry on with the half-delivered changes I have started last summer (including the docs).

**Comment by fredo-dedup** [comment]
Hi @scidom 
I started quick small proposal by doing a rewrite of MCJob to have your opinion. It is in the devel branch. I made MCJob a concrete type with a symbol parameter signalling if it is plain or task. That should solve the problem of trying to have constructors on an abstract type.
I can proceed with the idea if you think that can work. I'll have a few questions next.

**Comment by scidom** [comment]
@fredo-dedup if I understand your suggestion correctly, you  mean that the abstract `MCJob` type can be omitted and instead use the synonymous  function for job construction? If so, that makes sense, I don't have strong feelings about the abstract type.

On a second note, I wanted to ask you if it would be easy to provide a high-level description of the DSL you have in mind for model specification.

**Comment by fredo-dedup** [comment]
Presently I just want to upgrade Lora to the last version of ReverseDiffSource without user facing changes: it is still expression based and translates ~ into a log-likelihood accumulation term. The only changes would be that for loops are now allowed and that there can be higher order derivations. Once that is done we could start talking about how to best express the extensions  that you list at the top of this thread. My thinking is that bringing Lora to a usable state would make a useful first step.
To allow for future package architecture changes, I have also put the DSL code into a sub-module of Lora, this will make it easier to separate it from Lora if it is deemed preferable in the future.

As for the MCJob concept, I understand that it is the object that the user runs and resumes, `run` produces one or more chains and `resume` grows an existing chain or chains. It can contain several models (if you want to to have several chains with different starting values, or if you want to run sequential MC or serial tempering) or samplers (you want to benchmark different sampling methods, or different choices or sampling/tuning parameters for the same sampling method), but I don't see the use case for different runners. What could it be ? Besides SeqMC and SerialMC require themselves several Model (but only a single sampler). Wouldn't it be better to have a single runner by job ?

**Comment by scidom** [comment]
Thanks for the overview @fredo-dedup, I now see what you want to achieve. I agree that we also need to press on with reviving Lora, we need to bring it back to fully operational state asap.

As for your remark re the `runner` field of the various `MCJob` types, I fully acknowledge your remark. I  have spent hours reasoning both in favor and against a vector of runners; in fact for future reference, I had written down my thoughts in issue #67, which is solely dedicated in the job-runner structure that you raised. It is a long stream of thoughts, but if you read the issue, it will give you an idea of my background thinking on the matter.

In short, the main point I make at the end of #67 is that although the type of runner is conceptually unique, its parameters-fields may different for subchains, for example, in a population MCMC scheme. Of course, current developments in MCMC haven't accounted for such case as of yet, but I have been thinking in the most general way possible. So, the same argument that we have in mind with regards to have multiple models, to allow for differing field values of its model, holds in principle for the runner too.

In any case, there is also the matter of how we manage a Metropolis-within-Gibbs sampler as well, where one needs a runner regulating the Metropolis and a runner regulating the Gibbs sampler, but this is for the future; if we start thinking about this from now, we will end up not updating Lora asap, which is higher priority.

**Comment by fredo-dedup** [comment]
My bad, I only read through open issues !

**Comment by scidom** [comment]
Great, I will also focus on your suggestion, i.e. to finish the porting of the missing bits. While you focus on the DSL adaptation to the newest `ReverseDiffSource`, I will focus on the porting of the remaining samplers and documentation. I have been struggling to find a bit of time lately, but I think I can squeeze some time again now.

After all this is done, then we can start thinking about more generic extensions to model specification so as to allow a broader spectrum of Gibbs related models.

**Comment by scidom** [comment]
This issue, and more specifically hierarchical models and Gibbs sampling, can be accommodated by the new graph-based model. The infrastructure is there, all I need to do is to define a new job type to accommodate Gibbs sampling. For this reason, I will close this issue (there is a separate issue for Gibbs sampling).

As for writing a parser/DSL sitting atop of the current package structure, I would like to postpone this decision making for later, after the new package functionality has been completed and consolidated. Then, on the basis of users' feedback, we can decide whether a DSL is necessary.
